### ISSUE TITLE
Undesirable type argument inference from return position

### ISSUE THREAD

**Issue by DanielRosenwasser** [Microsoft/TypeScript#20681]
```ts
let x: number[] = Array.from(
    [true, true, false]
)
```

Today, this code type checks because `Array.from<T, U>` takes an optional `mapfn` of type `(el: T, index: number) => U`.

When `U` is not inferred from the `mapfn`, it is instead inferred from the contextual type.

**Comment by DanielRosenwasser** [comment]
Looks like this is a partial duplicate of #19682 (and was fixed in the signature of `Array.from` ), but isn't this indicative of a broader problem?

**Comment by sandersn** [comment]
Yes:

```ts
declare function f<T, U = T>(xs: T[], g?: (x: T) => U): U[];
const ns: number[] = f([true]) //should error here: 'boolean' is not assignable to 'number'
```

The rules for defaults may specify that inference works this way currently, but I believe we decided change those rules in a recent design meeting.

**Comment by ajafff** [comment]
This was reported a few times as a bug in one of TSLint's rules lately. Ref: https://github.com/palantir/tslint/issues/3540 and duplicates

Would be awesome if this get's fixed.

**Comment by DanielRosenwasser** [comment]
Spoke to @RyanCavanaugh, seems reasonable to say that this should really be two overloads, but it is still pretty undesirable.
